src/paho/mqtt/client.py

```diff
@@ -2671,6 +2694,89 @@ def _reconnect_wait(self):
             time.sleep(min(remaining, 1))
             remaining = target_time - time_func()
 
+    def _get_proxies(self):
+        def proxy_is_valid(p):
+            if isinstance(p, dict):
+                return bool(p.get("proxy_type"))
+            elif isinstance(p, (list, tuple)):
+                return len(p) == 6 and bool(p[0])
+            else:
+                return False
+
+        proxies = []  # proxy servers to use in order of preference
+
+        if proxy_is_valid(self._proxy):
+            proxies.append(self._proxy)
+
+        socks_default = socks.get_default_proxy()
+        if proxy_is_valid(socks_default):
+            proxies.append({
+                "proxy_type": socks_default[0],
+                "proxy_addr": socks_default[1],
+                "proxy_port": socks_default[2],
+                "proxy_rdns": socks_default[3],
+                "proxy_username": socks_default[4],
+                "proxy_password": socks_default[5]
+            })
+
+        # Finally, look at environment variables ending in _proxy as long as the
+        # host we're trying to connect to isn't listed under the no_proxy
+        # environment variable (matches built-in module urllib's behavior)
+        if not urllib_dot_request.proxy_bypass(self._host):
```

Comment:
I can't find documentation on proxy_bypass. I'm unsure how safe it's to rely this undocumented method.
I'm especially thinking about:
* Since it's not documented, I don't known from which version of Python is appear
* Since it's not documented, why it can't go away in next Python version ?

Maybe at least adding an hasattr() would help (on the other hand it means that if the function disappear this feature will be silently ignored)

Reply:
Good point. How about:

```python
if not (hasattr(urllib_dot_request, "proxy_bypass") and urllib_dot_request.proxy_bypass(self._host)):
```

I think this makes sense as a fallback if the function disappears—since the function tells us which connections we *shouldn't* proxy, if the function doesn't exist, we proxy connections using the supplied connections by default.